### PR TITLE
Enhance msb4064 unexpected task attribute error

### LINKED ISSUE #3922: MSB4064 could have additional information
MSB4064 is fired when a task has been successfully loaded but is missing a field that the XML invocation of the task defined. This often arises due to versioning problems: a task is revised to add a parameter, the target that invokes it is changed to pass it, but at runtime the old version of the task is loaded (see #663 for common reasons why).

The error is

```
error MSB4064: The "NullableReferenceTypes" parameter is not supported by the "Csc" task. Verify the parameter exists on the task, and it is a settable public instance property.
```

This error could be more actionable if it dumped some more information, like assembly identity of the loaded task, and path on disk it was loaded from. Something like

```
error MSB4064: The "{propertyName}" parameter is not supported by the "{typeName}" task loaded from {assemblyIdentity} at path {path}. Verify that the UsingTask points to the correct assembly, the parameter exists on the task, and it is a settable public instance property.
```

### PR DESCRIPTION
**Pull request by BartoszKlonowski**
This pull request fixes #3922.

The *MSB4064* error has been provided with additional information about the assembly such as it's version, full name, etc.

---

Changes provided within this commit contains:
* modify the MSB4064 string in resources,
* adjust it's usage in the code throwing this error,
* run the default translations by compilation

---

Results:

The example of MSB4064 printed is presented below:
obraz

### CODE REVIEW

**Review by rainersigwald — COMMENTED** [review]
This looks great, but I have a vague recollection that it isn't enough. I'll try to figure out _why_ I think that and get back to you. Hopefully I can convince myself that past-me was just wrong!

**Comment by BartoszKlonowski** [comment]
@rainersigwald  Perhaps what you feel that is missing here is a path (on local disk) to the assembly?
Thing is, that in my first approach to this, I've tried to use:
```
_taskFactoryWrapper.TaskFactoryLoadedType.Assembly.AssemblyName
```
but when debugging it, this path has always been `null`.
So due to the fact that I couldn't find any better path-containing property/variable, I decided to not include the path in the error message, so there's no confusion in case of that path didn't really work for some user.
I've tried to avoid the possible situation, where the error message would just display the empty space instead of correct path.

**Comment by rainersigwald** [comment]
> Perhaps what you feel that is missing here is a path (on local disk) to the assembly?
> Thing is, that in my first approach to this, I've tried to use:
> 
> ```
> _taskFactoryWrapper.TaskFactoryLoadedType.Assembly.AssemblyLocation
> ```
> 
> but when debugging it, this path has always been `null`.
> So due to the fact that I couldn't find any better path-containing property/variable, I decided to not include the path in the error message, so there's no confusion in case of that path didn't really work for some user.

This seems pretty plausible! It would definitely be nice to have the path (I'd strongly prefer that over assembly identity) but I agree with your tradeoffs, and this is better than nothing.

Is that path `null` when you try with the minimal repro project I (just) shared in https://github.com/dotnet/msbuild/issues/3922#issuecomment-747095060?

**Comment by BartoszKlonowski** [comment]
@rainersigwald Answering your question: no, the path isn't null when reproducing the error with the example you've provided me with.
I've pushed the changes and updated the PR's description - please check.

**Comment by Forgind** [comment]
It looks like LoadedAssembly is null for the ValidateNonExistantParameter (sic) test. It looks artificial to me, but are there any circumstances when it would be executing a task without loading an assembly? If so, maybe add a second error possibility depending on whether that exists?

**Comment by cdmihai** [comment]
Looking through the code, `LoadedType.LoadedAssembly` is always checked for null, so apparently it can be null. My assumption is that tasks that are not loaded from assemblies have it set to null (like inline tasks for example). So you'll have to treat the null case somehow. Most basic is to revert to the old message if it's null. Another implementation is to revert to another message that writes out the task factory type, which hopefully should be enough to let the user guess where that task is coming from. The common case is to have assembly based tasks.

Regarding tests, you have to update the mocked types to have an assembly, probably over here: https://github.com/dotnet/msbuild/blob/27981f15c4a62661691281ee52d3c5b7b7793068/src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs#L1159-L1166

Then you need to write two tests: one when the assembly is null, and one when the assembly is not null, and assert the proper error message in both.

**Comment by Forgind** [comment]
@BartoszKlonowski, status update?

**Comment by BartoszKlonowski** [comment]
@Forgind Thanks for asking, and sorry for such a long delay...
No update from my side - Unfortunatelly I have no time to work on this.
If you're asking because you'd like to take it, then let me know if you have any questions. Otherwise, I'll try to push that forward by the end of this week.
Let me know.

**Comment by Forgind** [comment]
No worries! If you didn't want to work on it anymore, I was willing to try to get it ship-shape, since I think it's pretty close. End of this week (or end of next week) is fine. Thanks!

**Review by Forgind — APPROVED** [review]
LGTM! Thank you!

File: src/Build/Resources/Strings.resx [hunk 1]
```diff
@@ -1225,7 +1225,7 @@
     <comment>{StrBegin="MSB4091: "}</comment>
   </data>
   <data name="UnexpectedTaskAttribute" xml:space="preserve">
-    <value>MSB4064: The "{0}" parameter is not supported by the "{1}" task. Verify the parameter exists on the task, and it is a settable public instance property.</value>
+    <value>MSB4064: The "{0}" parameter is not supported by the "{1}" task loaded from assembly: {2} from the path: {3}. Verify that the parameter exists on the task, the UsingTask points to the correct assembly and it is a settable public instance property.</value>
```
Comment (ladipro) [inline src/Build/Resources/Strings.resx:1228]: super-nit: In many other error messages mentioning `UsingTask` I see it wrapped in angle brackets making it clear that it refers to an XML element. Also, a comma before the last `and` would be more consistent with the style of the rest of this file.

**Review by ladipro — APPROVED** [review]
Looks great! One small nit inline.

**Comment by Forgind** [comment]
Thanks @BartoszKlonowski!
